internal/ethapi/api.go

```diff
@@ -996,14 +995,110 @@ func (s *PublicBlockChainAPI) Call(ctx context.Context, args CallArgs, blockNrOr
 	return (hexutil.Bytes)(result), err
 }
 
+// Optimism note: The gasPrice in Optimism is modified to always return 1 gwei. We
+// use the gasLimit field to communicate the entire user fee. This is done for
+// for compatibility reasons with the existing Ethereum toolchain, so that the user
+// fees can compensate for the additional costs the sequencer pays for publishing the
+// transaction calldata
 func DoEstimateGas(ctx context.Context, b Backend, args CallArgs, blockNrOrHash rpc.BlockNumberOrHash, gasCap *big.Int) (hexutil.Uint64, error) {
-	// Retrieve the block to act as the gas ceiling
-	block, err := b.BlockByNumberOrHash(ctx, blockNrOrHash)
+	// 1a. fetch the data price, depends on how the sequencer has chosen to update their values based on the
+	// l1 gas prices
+	dataPrice, err := b.SuggestDataPrice(ctx)
 	if err != nil {
 		return 0, err
 	}
-	// For now always return the gas limit
-	return hexutil.Uint64(block.GasLimit() - 1), nil
+
```

Comment:
What do you think about doing a quick check that `args.Data` is not `nil`?

Reply:
See https://github.com/ethereum-optimism/go-ethereum/pull/273/files#diff-c426ecd2f7d247753b9ea8c1cc003f21fa412661c1f967d203d4edf8163da344R1046

Reply:
This check is not done in the original `DoEstimateGas` of geth so I don't see why we need to add that